tiledb/sm/metadata/metadata.h

```diff
@@ -0,0 +1,240 @@
+/**
+ * @file   metadata.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2019 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ This file defines class Metadata.
+ */
+
+#ifndef TILEDB_METADATA_H
+#define TILEDB_METADATA_H
+
+#include <map>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "tiledb/sm/buffer/buffer.h"
+#include "tiledb/sm/buffer/const_buffer.h"
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/misc/status.h"
+#include "tiledb/sm/misc/uri.h"
+
+namespace tiledb {
+namespace sm {
+
+/**
+ * Handles writing/reading metadata. The persistent format of the metadata is:
+ * `0 (char) | key_len (uint32_t) | key (char*) | value_type (Datatype) |
+ *  value_num (uint32_t) | values (void*)`
+ * or
+ * `1 (char) | key_len (uint32_t) | key (char*) | value_type (Datatype) |
```

Comment:
should this be `1 (char)|||||`?

Reply:
Sorry, this is dead code, I will remove. Thanks for catching it.

Reply:
I think having the representation here is useful, just if I understood correctly the representation when the value of `del` is `1` would be a single `char`.

Reply:
The representation is now visible through the `MetadataValue` struct.